// object-js/dist/core/Obj.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }(); //-------------------------------------------------------------------------------
// Imports
//-------------------------------------------------------------------------------

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _util = require('../util');

var _fns = require('./fns');

var _interfaces = require('./interfaces');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

//-------------------------------------------------------------------------------
// Declare Class
//-------------------------------------------------------------------------------

/**
 * @class
 * @implements {IClone}
 * @implements {IEquals}
 * @implements {IHashCode}
 */

var Obj = function () {

  //-------------------------------------------------------------------------------
  // Constructor
  //-------------------------------------------------------------------------------

  /**
   * @constructs
   */

  function Obj() {
    _classCallCheck(this, Obj);

    Object.defineProperty(this, '_hashCode', {
      value: null,
      writable: true,
      enumerable: false,
      configurable: false
    });
    this.generateInternalId();
  }

  //-------------------------------------------------------------------------------
  // Getters and Setters
  //-------------------------------------------------------------------------------

  /**
   * @return {number}
   */


  _createClass(Obj, [{
    key: 'getInternalId',
    value: function getInternalId() {
      if (!this._internalId) {
        this.generateInternalId();
      }
      return this._internalId;
    }

    //-------------------------------------------------------------------------------
    // IClone Implementation
    //-------------------------------------------------------------------------------

    /**
     * @param {boolean=} deep
     * @return {*}
     */

  }, {
    key: 'clone',
    value: function clone(deep) {
      var constructor = this.constructor;
      var cloneObject = new constructor();
      _lodash2.default.forIn(this, function (key, value) {
        if (!_lodash2.default.isFunction(value)) {
          if (deep) {
            cloneObject[key] = Obj.clone(value, deep);
          } else {
            cloneObject[key] = value;
          }
        }
      });
      return cloneObject;
    }

    //-------------------------------------------------------------------------------
    // IEquals Implementation
    //-------------------------------------------------------------------------------

    /**
     * If two Objs are equal then they MUST return the same hashCode. Otherwise the world will end!
     * @param {*} value
     * @return {boolean}
     */

  }, {
    key: 'equals',
    value: function equals(value) {
      return _lodash2.default.get(value, '_internalId') === this.getInternalId();
    }

    //-------------------------------------------------------------------------------
    // IHashCode Implementation
    //-------------------------------------------------------------------------------

    /**
     * Equal hash codes do not necessarily guarantee equality.
     * @return {number}
     */

  }, {
    key: 'hashCode',
    value: function hashCode() {
      if (!this._hashCode) {
        this._hashCode = (0, _util.hash)(this);
      }
      return this._hashCode;
    }

    //-------------------------------------------------------------------------------
    // Protected Methods
    //-------------------------------------------------------------------------------

    /**
     * @protected
     */

  }, {
    key: 'generateInternalId',
    value: function generateInternalId() {
      (0, _util.ensureId)(this);
    }
  }]);

  return Obj;
}();

//-------------------------------------------------------------------------------
// Interfaces
//-------------------------------------------------------------------------------

exports.default = Obj;
(0, _fns.implement)(Obj, _interfaces.IClone);
(0, _fns.implement)(Obj, _interfaces.IEquals);
(0, _fns.implement)(Obj, _interfaces.IHashCode);
