// object-js/dist/core/fns/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _clone = require('./clone');

Object.defineProperty(exports, 'clone', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_clone).default;
  }
});

var _doesExtend = require('./doesExtend');

Object.defineProperty(exports, 'doesExtend', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_doesExtend).default;
  }
});

var _doesImplement = require('./doesImplement');

Object.defineProperty(exports, 'doesImplement', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_doesImplement).default;
  }
});

var _equals = require('./equals');

Object.defineProperty(exports, 'equals', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_equals).default;
  }
});

var _hashCode = require('./hashCode');

Object.defineProperty(exports, 'hashCode', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_hashCode).default;
  }
});

var _implement = require('./implement');

Object.defineProperty(exports, 'implement', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_implement).default;
  }
});

var _newInstance = require('./newInstance');

Object.defineProperty(exports, 'newInstance', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_newInstance).default;
  }
});

var _newInstanceWithArray = require('./newInstanceWithArray');

Object.defineProperty(exports, 'newInstanceWithArray', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_newInstanceWithArray).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
